serverless/experimental/wasm/main.go

```diff
@@ -0,0 +1,219 @@
+		if err != nil {
+			panic(fmt.Errorf("Failed to integrate: %q", err))
+		}
+		if newCp == nil {
+			logMsg("Nothing to integrate")
+			return nil
+		}
+		if err != nil {
+			logMsg(err)
+			return nil
+		}
+
+		newCp.Ecosystem = ecosystem
+		cpRaw := newCp.Marshal()
+		if err := logStorage.WriteCheckpoint(cpRaw); err != nil {
+			logMsg(err)
+			return nil
+		}
+
+		logMsg("Integrate OK")
+		showCP()
+		return nil
+	})
+	return jsonFunc
+}
+func sequence() js.Func {
+	jsonFunc := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		if len(args) != 0 {
+			return "Invalid number of arguments passed - want no args"
+		}
+
+		pendingLeaves, err := logStorage.PendingKeys()
+		if err != nil {
+			logMsg(err)
+			return nil
+		}
+		for _, lk := range pendingLeaves {
+			l, err := logStorage.Pending(lk)
+			if err != nil {
+				logMsg(err)
+				return nil
+			}
+			h := rfc6962.DefaultHasher.HashLeaf(l)
+			isDupe := false
+			seq, err := logStorage.Sequence(h, l)
+			if err != nil {
+				if !errors.Is(err, storage.ErrDupeLeaf) {
+					logMsg(err)
+					return nil
+				}
+				isDupe = true
+			}
+			s := fmt.Sprintf("index %d: %v", seq, lk)
+			if isDupe {
+				s += " (dupe)"
+			}
+			logMsg(s)
+			if err := logStorage.DeletePending(lk); err != nil {
+				logMsg(err)
+				return nil
+			}
+
+		}
+		return nil
+	})
+	return jsonFunc
+}
+func queueLeaf() js.Func {
+	jsonFunc := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		if len(args) != 1 {
+			return "Invalid number of arguments passed - want 1 leaf"
+		}
+		leaf := args[0].String()
+		fmt.Printf("leaf %q\n", leaf)
+
+		if err := logStorage.Queue([]byte(leaf)); err != nil {
+			return fmt.Sprintf("failed to queue leaf: %v", err)
+		}
+		return nil
+	})
+	return jsonFunc
+}
+
+func monitor() {
+	cpCur := &logfmt.Checkpoint{}
+
+	for {
+		<-time.Tick(time.Second)
+		cpRaw, _ := webstorage.ReadCheckpoint(logPrefix)
+		cp := &logfmt.Checkpoint{}
+		if _, err := cp.Unmarshal(cpRaw); err != nil {
+			monMsg(fmt.Sprintf("Couldn't parse CP: %v", err))
+			continue
+		}
+
+		if cp.Size > cpCur.Size {
+			monMsg("----------------------------------")
+			monMsg(fmt.Sprintf("Saw new CP:\n%s", string(cp.Marshal())))
+			cpCur = cp
+		}
+
+	}
+}
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	fmt.Println("Serverless Web Assembly!")
+	js.Global().Set("queueLeaf", queueLeaf())
+	js.Global().Set("sequence", sequence())
+	js.Global().Set("integrate", integrate())
+
+	cpRaw, err := webstorage.ReadCheckpoint(logPrefix)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			panic(err)
+		}
+		logStorage, err = webstorage.Create(logPrefix, []byte("empty"))
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		cp := &logfmt.Checkpoint{}
+		_, err := cp.Unmarshal(cpRaw)
+		if err != nil {
+			panic(err)
+		}
+		logStorage, err = webstorage.Load(logPrefix, cp)
+		if err != nil {
+			panic(err)
+		}
+	}
+	showCP()
+	go monitor()
+	<-make(chan bool)
```

Comment:
This is odd. Is this a wasm idiom?

Reply:
I guess it's not unlike the `http.ListenAndServe` thing - if you don't do this the program exits (in the WASM case that's bad since the JS world can call into the funcs it's published)